multus/multus.go

```diff
@@ -415,25 +424,24 @@ func cmdDel(args *skel.CmdArgs, exec invoke.Exec, kubeClient k8s.KubeClient) err
 		return logging.Errorf("Multus: Err in getting k8s args: %v", err)
 	}
 
-	numK8sDelegates, kc, err := k8s.TryLoadPodDelegates(k8sArgs, in, kubeClient)
+	_, kc, err := k8s.TryLoadPodDelegates(k8sArgs, in, kubeClient)
 	if err != nil {
 		return err
 	}
 
-	if numK8sDelegates == 0 {
-		// re-read the scratch multus config if we have only Multus delegates
-		netconfBytes, err := consumeScratchNetConf(args.ContainerID, in.CNIDir)
-		if err != nil {
-			if os.IsNotExist(err) {
-				// Per spec should ignore error if resources are missing / already removed
-				return nil
-			}
-			return logging.Errorf("Multus: Err in  reading the delegates: %v", err)
+	// Read the cache to get delegates json for the pod
+	netconfBytes, path, err := consumeScratchNetConf(args.ContainerID, in.CNIDir)
+	defer os.Remove(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// Per spec should ignore error if resources are missing / already removed
```

Comment:
If we're going to always use the cached delegates we need to make sure that the cache dir is persistent on-disk, eg that the DaemonSet uses a volume mount to persist the cache on the host.

But now we have an upgrade problem. When somebody updates from old multus to this one while the cluster is running (which is very common in OpenShift 4 for example), there won't necessarily be cache files. Then when kubelet deletes the pod, the delete will fail because the cache is empty.

Per https://github.com/intel/multus-cni/pull/246#pullrequestreview-201630484 I think we still do need to fall back to the K8s delegates if the cache is empty. While not ideal, that will fix both the upgrade case and possibly user errors.

So perhaps move the TryLoadPodDelegates() after the consumeScratchNetConf(), and if if os.IsNotExist(err) is true then we use the delegates from TryLoadPodDelegates()?

Reply:
@s1061123 ^

Reply:
@dcbw , thank you for your comment! Now I'm modifying and will update the diff soon.

Reply:
Update the diffs with the fix.